### PR TITLE
Implemented experimental corenet application protocol support.

### PR DESCRIPTION
**Pull request by tswindell**
Signed-off-by: Tom Swindell <t.swindell@rubyx.co.uk>

### CODE REVIEW

**Comment by tswindell** [comment]
This is a replacement for https://github.com/ipfs/go-ipfs/pull/3128

File: test/sharness/t0180-corenet.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  ipfsi 1 exp corenet dial $PEERID_0 corenet-test /ip4/127.0.0.1/tcp/10002 2>&1 > dialer-stdouterr.log
+
+  nc -v 127.0.0.1 10002 | dd of=client.out
+'
+
+test_expect_success 'Test server to client communications' '
+  nc -l -p 10001 | dd of=server.out &
+
+  ipfsi 1 exp corenet dial $PEERID_0 corenet-test /ip4/127.0.0.1/tcp/10002 2>&1 > dialer-stdouterr.log
```
Comment (Kubuxu) [inline test/sharness/t0180-corenet.sh]: Connect this and the next line with `&&`.

File: test/sharness/t0180-corenet.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+test_expect_success 'Test server to client communications' '
+  dd if=corenet0.bin | nc -l -p 10001 &
+
+  ipfsi 1 exp corenet dial $PEERID_0 corenet-test /ip4/127.0.0.1/tcp/10002 2>&1 > dialer-stdouterr.log
+
+  nc -v 127.0.0.1 10002 | dd of=client.out
+'
+
+test_expect_success 'Test server to client communications' '
+  nc -l -p 10001 | dd of=server.out &
```
Comment (Kubuxu) [inline test/sharness/t0180-corenet.sh]: Always collect PID of background jobs to: 1. check their return code, 2. try killing them when tests are done.

File: test/sharness/t0180-corenet.sh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+'
+
+test_expect_success 'start ipfs listener' '
+  ipfsi 0 exp corenet listen corenet-test /ip4/127.0.0.1/tcp/10001 2>&1 > listener-stdouterr.log
+'
+
+test_expect_success 'Test server to client communications' '
+  dd if=corenet0.bin | nc -l -p 10001 &
+
+  ipfsi 1 exp corenet dial $PEERID_0 corenet-test /ip4/127.0.0.1/tcp/10002 2>&1 > dialer-stdouterr.log
```
Comment (tswindell) [inline test/sharness/t0180-corenet.sh]: Can't do that, because nc -l blocks until disconnect.

File: test/sharness/t0180-corenet.sh [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+'
+
+test_expect_success 'generate test data' '
+  dd if=/dev/urandom of=corenet0.bin bs=1K count=1 &&
+  dd if=/dev/urandom of=corenet1.bin bs=1K count=1
+'
+
+# netcat (nc) is needed for the following tests
+test_expect_success "nc is available" '
+  type nc >/dev/null
```
Comment (Kubuxu) [inline test/sharness/t0180-corenet.sh]: We use `nc` in other tests so it isn't required.

File: test/sharness/t0180-corenet.sh [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+startup_cluster 2
+
+test_expect_success 'peer ids' '
+  PEERID_0=$(iptb get id 0) &&
+  PEERID_1=$(iptb get id 1)
+'
+
+test_expect_success 'generate test data' '
+  dd if=/dev/urandom of=corenet0.bin bs=1K count=1 &&
```
Comment (Kubuxu) [inline test/sharness/t0180-corenet.sh]: Either use `go-random` or use some pre-set text data, easier to see on diff.

File: test/sharness/t0180-corenet.sh (hunk 3 above)
Comment (Kubuxu) [inline test/sharness/t0180-corenet.sh]: I mean line bellow that.

File: test/sharness/t0180-corenet.sh [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+test_expect_success "nc is available" '
+  type nc >/dev/null
+'
+
+test_expect_success 'start ipfs listener' '
+  ipfsi 0 exp corenet listen corenet-test /ip4/127.0.0.1/tcp/10001 2>&1 > listener-stdouterr.log
+'
+
+test_expect_success 'Test server to client communications' '
+  dd if=corenet0.bin | nc -l -p 10001 &
```
Comment (Kubuxu) [inline test/sharness/t0180-corenet.sh]: You can use just `<`, instead of dd.

File: test/sharness/t0180-corenet.sh [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  NC_SERVER_PID=$!
+
+  ipfsi 1 exp corenet dial $PEERID_0 corenet-test /ip4/127.0.0.1/tcp/10002 2>&1 > dialer-stdouterr.log &&
+  nc -v 127.0.0.1 10002 | dd of=client.out
+
+  wait $NC_SERVER_PID
+'
+
+test_expect_success 'Test server to client communications' '
+  nc -l -p 10001 | dd of=server.out &
```
Comment (Kubuxu) [inline test/sharness/t0180-corenet.sh]: same about `dd`.

**Comment by whyrusleeping** [comment]
I'm planning on reviewing and getting this ready to go hopefully for the 0.4.8 release. Since we're clear about the experimental nature of it i think that its fine to merge it without perfecting the interfaces.

**Comment by MichaelMure** [comment]
That would be very handy for my project. Currently I have to use pubsub, even for direct message between two nodes.

**Comment by whyrusleeping** [comment]
@MichaelMure we're going to be working on getting this ready to ship soon. @magik6k is going to be pushing to get it merged :)

**Comment by MichaelMure** [comment]
@magik6k, stranger of the internet,  I'm rooting for you :-p

**Comment by Kubuxu** [comment]
Implemented in #3943

**Pull request closed without merging**